rholang/src/main/scala/coop/rchain/rholang/interpreter/Registry.scala

```diff
@@ -0,0 +1,216 @@
+    replaceChan match {
+      case Channel(q @ Quote(_)) =>
+        handleResult(
+          space
+            .produce(q, ListChannelWithRandom(Seq(data), dataRand, None), false))
+      case _ => Task.unit
+    }
+
+  def failAndReplace(data: Channel,
+                     replaceChan: Channel,
+                     retChan: Channel,
+                     dataRand: Blake2b512Random,
+                     failRand: Blake2b512Random): Task[Unit] =
+    for {
+      _ <- replace(data, replaceChan, dataRand)
+      _ <- fail(retChan, failRand)
+    } yield ()
+
+  def fetchData(dataSource: Quote,
+                key: Channel,
+                ret: Channel,
+                rand: Blake2b512Random): Task[Unit] = {
+    val channel: Par = GPrivate(ByteString.copyFrom(rand.next()))
+    for {
+      _ <- handleResult(
+            space.produce(Quote(channel),
+                          ListChannelWithRandom(Seq(key, ret, Channel(dataSource)), rand, None),
+                          false))
+      _ <- handleResult(
+            space.consume(
+              Seq[Channel](Quote(channel), dataSource),
+              Seq(prefixRetReplacePattern, triePattern),
+              TaggedContinuation(ScalaBodyRef(lookupCallbackRef)),
+              false
+            ))
+    } yield ()
+  }
+
+  def lookup(args: RootSeq[ListChannelWithRandom]): Task[Unit] =
+    args match {
+      case Seq(ListChannelWithRandom(Seq(key, ret), rand, cost)) =>
+        key match {
+          case Channel(Quote(keyPar)) =>
+            keyPar.singleExpr match {
+              case Some(Expr(GByteArray(_))) =>
+                fetchData(Quote(registryRoot), key, ret, rand)
+              case _ =>
+                fail(ret, rand)
+            }
+          case _ =>
+            fail(ret, rand)
+        }
+      case _ => Task.unit
+    }
+
+  // A lookup result should result in 1 of 3 things
+  // Result not there. Return Nil
+  // Result there, return it.
+  // Further lookup needed, recurse.
+
+  def lookupCallback(args: RootSeq[ListChannelWithRandom]): Task[Unit] =
+    args match {
+      case Seq(ListChannelWithRandom(Seq(key, ret, replaceChan), callRand, callCost),
+               ListChannelWithRandom(Seq(data), dataRand, dataCost)) =>
+        def localFail() = failAndReplace(data, replaceChan, ret, dataRand, callRand)
+        key match {
+          case Channel(Quote(keyPar)) =>
+            keyPar.singleExpr match {
+              case Some(Expr(GByteArray(bs))) =>
+                val head = if (bs.isEmpty()) bs else bs.substring(0, 1)
+                val tail = if (bs.isEmpty()) bs else bs.substring(1)
+                data match {
+                  case Channel(Quote(dataPar)) =>
+                    dataPar.singleExpr match {
+                      case Some(Expr(EMapBody(parMap))) =>
+                        parMap.ps.get(Expr(GByteArray(head))) match {
+                          case None => localFail()
+                          case Some(value) =>
```

Comment:
How do you call localFail() in that case?
Also, I don't like writing code I don't understand, and there's magic going on that I don't understand. So you'll have to explain it better.

Reply:
> So you'll have to explain it better.

Every assignment in a for-comprehension is a pattern match, or a destructuring assignment to be more precise. The difference is: in a pattern match you get to decide what to do for each case; in a D.A. you just get the values assigned if everything matches or a MatchError otherwise.

Notice the `: Any`-s hiding the type info from the compiler in the example below:

```scala
val key: Any = Some(42)
val v1: Any  = EVar(Var(FreeVar(1)))
val v2: Any  = EVar(Var(FreeVar(2)))
val vs: Any  = v1 :: v2 :: Nil

for {
  _                         <- ().pure[Id] //Any Monad will do just fine
  Some(k)                   = key
  EVar(a) :: EVar(b) :: Nil = vs
  Var(FreeVar(x))           = a
  Var(FreeVar(y))           = b
} yield (k, x, y)

//prints:
//res0: cats.Id[(Any, Int, Int)] = (42,1,2)
```

> How do you call localFail() in that case?

If any of the patterns doesn't match, a MatchError is thrown. Since our handling of non-matches is always the same and doesn't rely on the things matched so far, we can simply:

```scala
try {
  for {
  //...
  } yield result
} catch {
  case x: MatchError => localFail()
}
```

I hate resorting to the `try/catch` too, but I'd say the readability gain justifies that.

Reply:
If not for the `Task` effect around, we could just do:
```scala
try {
  val Some(k)                   = key
  val EVar(a) :: EVar(b) :: Nil = vs
  val Var(FreeVar(x))           = a
  val Var(FreeVar(y))           = b
  (x, y, k) //return result
} catch {
  case x: MatchError => localFail()
}
```

Reply:
I understand the last thing you wrote.
I think we can do it in that style.